src/nvim/terminal.c

```diff
@@ -0,0 +1,1114 @@
+      changed++;
+    } else {
+      ml_append(linenr - 1, (uint8_t *)term->textbuf, 0, false);
+      added++;
+    }
+  }
+
+  int change_start = row_to_linenr(term, term->invalid_start);
+  int change_end = change_start + changed;
+  changed_lines(change_start, 0, change_end, added);
+  term->invalid_start = INT_MAX;
+  term->invalid_end = -1;
+}
+
+static void redraw(bool restore_cursor)
+{
+  int save_row, save_col;
+  if (restore_cursor) {
+    // save the current row/col to restore after updating screen when not
+    // focused
+    save_row = ui_current_row();
+    save_col = ui_current_col();
+  }
+  block_autocmds();
+  validate_cursor();
+
+  if (must_redraw) {
+    update_screen(0);
+  }
+
+  redraw_statuslines();
+
+  if (need_maketitle) {
+    maketitle();
+  }
+
+  showruler(false);
+
+  Terminal *term = curbuf->terminal;
+  if (restore_cursor) {
+    ui_cursor_goto(save_row, save_col);
+  } else if (term && is_focused(term)) {
+    curwin->w_wrow = term->cursor.row;
+    curwin->w_wcol = term->cursor.col + win_col_off(curwin);
+    setcursor();
+  }
+
+  unblock_autocmds();
+  ui_flush();
+}
+
+static void adjust_topline(Terminal *term)
+{
+  int height, width;
+  vterm_get_size(term->vt, &height, &width);
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    if (wp->w_buffer == term->buf) {
+      // for every window that displays a terminal, ensure the cursor is in a
+      // valid line
+      wp->w_cursor.lnum = MIN(wp->w_cursor.lnum, term->buf->b_ml.ml_line_count);
+      if (curbuf != term->buf || is_focused(term)) {
+        // if the terminal is not in the current window or if its focused,
+        // adjust topline/cursor so the window will "follow" the terminal
+        // output
+        wp->w_cursor.lnum = term->buf->b_ml.ml_line_count;
+        set_topline(wp, MAX(wp->w_cursor.lnum - height + 1, 1));
+      }
+    }
+  }
+}
+
+static int row_to_linenr(Terminal *term, int row)
+{
+  return row != INT_MAX ? row + (int)term->sb_current + 1 : INT_MAX;
+}
+
+static int linenr_to_row(Terminal *term, int linenr)
+{
+  return linenr - (int)term->sb_current - 1;
+}
+
+static bool is_focused(Terminal *term)
+{
+  return State & TERM_FOCUS && curbuf == term->buf;
+}
+
+#define GET_CONFIG_VALUE(t, k, o)                                        \
+  do {                                                                   \
+    Error err;                                                           \
+    o = dict_get_value(t->buf->b_vars, cstr_as_string(k), &err);         \
+    if (obj.type == kObjectTypeNil) {                                    \
+      o = dict_get_value(&globvardict, cstr_as_string(k), &err);         \
+    }                                                                    \
+  } while (0)
+
+static char *get_config_string(Terminal *term, char *key)
+{
+  Object obj = OBJECT_INIT;
+  GET_CONFIG_VALUE(term, key, obj);
+  if (obj.type == kObjectTypeString) {
+    return obj.data.string.data;
+  }
```

Comment:
Do we need to use `api_free_object(obj)` here and later free the returned pointer?